## A: For the given data write and run regex

### Q1. How many records have a date that is expressed without using alphabets?

In [19]:
import pandas as pd
import re
data = pd.read_csv("Data.csv")
data.head()

# Comprehensive regex pattern for dates
date_pattern = r"""
    # Match dates with slashes, dashes or dots
    (?:
        # MM/DD/YY or MM/DD/YYYY (American format)
        (?:0?[1-9]|1[0-2])[/\-\.](?:0?[1-9]|[12][0-9]|3[01])[/\-\.](?:19|20)?\d\d
        |
        # DD/MM/YY or DD/MM/YYYY (European format)
        (?:0?[1-9]|[12][0-9]|3[01])[/\-\.](?:0?[1-9]|1[0-2])[/\-\.](?:19|20)?\d\d
        |
        # YYYY/MM/DD (ISO format)
        (?:19|20)\d\d[/\-\.](?:0?[1-9]|1[0-2])[/\-\.](?:0?[1-9]|[12][0-9]|3[01])
        |
        # YYYYMMDD (Compact ISO format)
        (?:19|20)\d\d(?:0[1-9]|1[0-2])(?:0[1-9]|[12][0-9]|3[01])
    )
"""

# Compile the regex pattern first
compiled_pattern = re.compile(date_pattern, re.VERBOSE)

# Use a lambda function to apply the compiled pattern
date_mask = data['Data'].apply(lambda x: bool(compiled_pattern.search(str(x))) if pd.notna(x) else False)

# Show the count
date_count = date_mask.sum()
if date_count > 0:
    # Displaying all matching records
    date_records = data[date_mask]
    print("Records with dates (no alphabets):")
    display(date_records)
    print("\n")
print("Total number of records with dates expressed without alphabets:", date_count)

Total number of records with dates expressed without alphabets: 0


**There is no records with just numeric dates**

## B: Perform stemming and lemmatization

### Q1. Use porter stemmer to run stemming. Count the number of unique words/tokens before and after stemming

In [20]:
import pandas as pd
import re
import nltk
from nltk.stem import PorterStemmer
from nltk.tokenize import word_tokenize
from collections import Counter

# Download necessary NLTK resources
nltk.download('punkt')

# Initialize the Porter stemmer
porter = PorterStemmer()

# Function to clean and tokenize text
def clean_and_tokenize(text):
    if not isinstance(text, str):
        return []
    
    # Convert to lowercase and remove special characters
    text = text.lower()
    text = re.sub(r'[^a-zA-Z0-9\s]', '', text)
    
    # Tokenize
    tokens = word_tokenize(text)
    
    # Remove very short words (optional)
    tokens = [word for word in tokens if len(word) > 1]
    
    return tokens

# Function to stem tokens
def stem_tokens(tokens):
    return [porter.stem(token) for token in tokens]

# Process the entire dataset
all_tokens_before = []
all_tokens_after = []

# Apply tokenization and stemming to each row
for text in data['Data']:
    tokens = clean_and_tokenize(text)
    all_tokens_before.extend(tokens)
    
    stemmed_tokens = stem_tokens(tokens)
    all_tokens_after.extend(stemmed_tokens)

# Count unique tokens before and after stemming
unique_before = len(set(all_tokens_before))
unique_after = len(set(all_tokens_after))

# Get the total number of tokens
total_tokens = len(all_tokens_before)

# Print results
print(f"Total number of tokens: {total_tokens}")
print(f"Number of unique tokens before stemming: {unique_before}")
print(f"Number of unique tokens after stemming: {unique_after}")
print(f"Reduction in vocabulary size: {unique_before - unique_after} words ({((unique_before - unique_after) / unique_before * 100):.2f}%)")

# Optional: Show the most common words before and after stemming
print("\nMost common words before stemming:")
print(Counter(all_tokens_before).most_common(10))

print("\nMost common words after stemming:")
print(Counter(all_tokens_after).most_common(10))

[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\Admin\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!


Total number of tokens: 110966
Number of unique tokens before stemming: 8109
Number of unique tokens after stemming: 6752
Reduction in vocabulary size: 1357 words (16.73%)

Most common words before stemming:
[('and', 3516), ('the', 2859), ('of', 2223), ('to', 2179), ('in', 1539), ('for', 1507), ('alberta', 952), ('is', 922), ('my', 895), ('on', 827)]

Most common words after stemming:
[('and', 3516), ('the', 2859), ('of', 2223), ('to', 2179), ('in', 1545), ('for', 1507), ('alberta', 1061), ('is', 922), ('my', 895), ('on', 827)]


**Stemming Analysis using Porter Stemmer**
- Total number of tokens: **110,966**
- Number of unique tokens before stemming: **8,109**
- Number of unique tokens after stemming: **6,752**
- Reduction in vocabulary size: **1,357 words** (**16.73%** decrease)

## Reference

- Claude AI Prompt 1: Python regular expression for all formats of numeric dates
- Claude AI final prompt: UserWarning: This pattern is interpreted as a regular expression, and has match groups. To actually get the groups, use str.extract.
  date_mask = data['Data'].str.contains(date_pattern, regex=True, flags=re.VERBOSE)
- Claude AI prompt : Given the data info how to perform stemming with porter stemmer.